### PR TITLE
Coverity

### PR DESCRIPTION
**Pull request by gregory38**
Warning/Error reports by coverity. Please review the change.

### CODE REVIEW

**Comment by avih** [comment]
There are not too many changes, and while some of them are relatively obvious, for others I don't see clearly the reason. Could you please attach someplace the coverity output log which references the issues modified by this PR?

**Comment by refractionpcsx2** [comment]
The changes look like they wouldn't cause any problems, but it would be nice to know what the warnings were that it gave.  As I said on the forum, there's a lot of places we know what to expect so it doesn't really matter so some are a little pointless. But if we're going to make it look like our code is cool and well done then I guess they need to be done.

**Comment by gregory38** [comment]
Try this link. Normally you can login with your github account
https://scan.coverity.com/projects/pcsx2-pcsx2
Most issue are related to the level high. Check submitted issue. You could check filename.

I agree that some fixes are "pointless". However they might create issue on the future. And they just create noise for the real issues.

I didn't fixes all coverity Error/Warning. Lots of them remains but it takes some times.

**Comment by avih** [comment]
I generally dislike registering new services which I don't intend to use frequently, even if it's via the github account.

Could you please post the log? Or even better, put it at the commit message such that it's attached to the actual patches.

**Comment by gregory38** [comment]
Technically you don't register, you just login as you login in github. Nothing more.

There isn't really a log. It is web2.0, so you have a nice interface that highlight the issue directly in the log with the explanation of all the required condition that lead to the potential issue.

**Comment by gregory38** [comment]
Tell me which fixes require (hopefully not too much) more info and I will post them.

**Comment by avih** [comment]
TBH, like I said earlier, It would have been best IMO if every commit message included the coverity issue output for the changes on that commit. This way the reasons for the changes would be part of the git repo itself.

**Comment by gregory38** [comment]
Yes but I don't have a nice report. (Or I miss it). I could copy the type of defect but it won't be better as the actual commit message. I think the intent of the change is obvious. However we could debate that some of the changes won't have any functional impact.
For example, the 2nd commit add a free of the index. By contruction (well it seems) len will be 0 if index is NULL so no need to clean it. However if someone changes the gzip function, it might introduce a mem leak. The new behavior is only safer.

**Comment by gregory38** [comment]
I updated commit number 8 (debugger: enforce a null terminated char of a C-string) based on turteli feedback.
I pushed 2 new additional commits that protect 2 array indexes to avoid out-of-bound access.

**Comment by gregory38** [comment]
New update that fix rather important issue. Operation overflow/infinite loop/wrong condition on the interpreter.

Remaining coveritiy issues are less important so I will do a separate PR for them. It is already big enough.

**Comment by ssakash** [comment]
In my opinion, It'd be better to deal with them in a single PR rather than Multiple ones. this way all of these coverity commits could be reviewed at a single place.

**Comment by gregory38** [comment]
There are already 20 commits to review. It seems enough for me. Besides I need some times to analye coverity report and to fix remaining issues.

**Comment by micove** [comment]
Visual Studio has the Code Analyzer which found some of these issues or at least I remember it complaining about possible null dereference in GSdx. Also valgrind reported some of these too. These can be used if using coverity is too much of a pain. Parsing through the sea of complaints from both of these tools is another matter.

**Comment by gregory38** [comment]
Personally I find coverity quite nice. Try the link above.
It tries to explain how the issue has been found and tell you the hypothesis. You can see the inline definition/declaration of the context around. You can sort/filter the report in various way.

It would be nice to run others tool. There is an issue with PVS report.

**Comment by gregory38** [comment]
Could the branch be merged? Or is anyone want to review/comment some commits.

File: pcsx2/x86/ix86-32/iR5900-32.cpp [hunk 1]
```diff
@@ -710,7 +710,7 @@ static void recResetRaw()
 
 	maxrecmem = 0;
 
-	memset(recConstBuf, 0, RECCONSTBUF_SIZE * sizeof(recConstBuf));
+	memset(recConstBuf, 0, RECCONSTBUF_SIZE * sizeof(*recConstBuf));
```
Comment (avih) [inline pcsx2/x86/ix86-32/iR5900-32.cpp]: This seems pretty serious. Do we know how/where it affected us?

File: plugins/GSdx/GSState.cpp [hunk 2]
```diff
@@ -5252,7 +5252,7 @@ bool GSC_UrbanReign(const GSFrameInfo& fi, int& skip)
 {
 	if(skip == 0)
 	{
-		if(fi.TME && fi.FBP==0x0000 && fi.TBP0==0x3980 && fi.FPSM==fi.TPSM && fi.TPSM == PSM_PSMCT32 && fi.TPSM ==0  && fi.FBMSK == 0x0)
+		if(fi.TME && fi.FBP==0x0000 && fi.TBP0==0x3980 && fi.FPSM==fi.TPSM && fi.TPSM == PSM_PSMCT32 && fi.FBMSK == 0x0)
```
Comment (avih) [inline plugins/GSdx/GSState.cpp]: How is this related to Coverity?

**Comment by avih** [comment]
Other than my two questions above, I'm good with it.

File: plugins/GSdx/GS.cpp [hunk 3]
```diff
@@ -1670,6 +1670,8 @@ EXPORT_C GSReplay(char* lpszCmdLine, int renderer)
 			sleep(1);
 		} else {
 			unsigned long end = timeGetTime();
+			frame_number = std::max(1ul, frame_number); // avoid a potential division by 0
+
 			fprintf(stderr, "The %ld frames of the scene was render on %ldms\n", frame_number, end - start);
```
Comment (ssakash) [inline plugins/GSdx/GS.cpp]: was render -> was rendered 

I know this not relevant to the PR, just spotted some small grammatical mistake.

File: pcsx2/x86/ix86-32/iR5900-32.cpp (hunk 1 above)
Comment (gregory38) [inline pcsx2/x86/ix86-32/iR5900-32.cpp]: If I trust coverity report, there is no impact. The type is the same size as the pointer. However it will crash on 64 bits.

File: plugins/GSdx/GSState.cpp (hunk 2 above)
Comment (gregory38) [inline plugins/GSdx/GSState.cpp]: Coverity reports dumb conditions. In this case TPSM was tested against 2 values. 0 and PSMCT32 (strangely coverity saw them as not egual)

**Comment by turtleli** [comment]
Reviewed, I think it's ok.

**Comment by gregory38** [comment]
Ok. Great. Thanks all for the review.
